File: examples/hello-world.ts
/**
 * Manual smoke test for widevin.
 *
 * Run:
 *   pnpm run example:hello
 *   pnpm run example:hello -- "Explain what this library does in one sentence."
 *
 * First run opens your browser to sign in to Devin/Cascade. The resulting
 * token is cached in `.devin-token` at the repo root (gitignored), so later
 * runs skip the browser step. Set DEVIN_TOKEN to use an existing token
 * instead of logging in.
 *
 * This file imports from `../src/index.js` because it lives inside the
 * widevin repo itself. A real consumer would instead write:
 *
 *   import { createDevinProvider, createFileTokenStore } from "widevin";
 */
import { spawn } from "node:child_process";
import {
  createDevinProvider,
  createFileTokenStore,
  DevinApiError,
  DevinAuthError,
  DevinProtocolError
} from "../src/index.js";

const TOKEN_PATH = ".devin-token";

const openUrlInBrowser = (url: string): void => {
  console.error(`Open this URL to sign in to Devin: ${url}`);
  const [command, args] =
    process.platform === "darwin"
      ? (["open", [url]] as const)
      : process.platform === "win32"
        ? (["cmd", ["/c", "start", "", url]] as const)
        : (["xdg-open", [url]] as const);
  try {
    spawn(command, args, { stdio: "ignore", detached: true }).unref();
  } catch {
    console.error("Could not launch a browser automatically; open the URL above manually.");
  }
};

const main = async (): Promise<void> => {
  const message = process.argv.slice(2).join(" ") || "Say hello in one short sentence.";
  const tokenStore = createFileTokenStore(TOKEN_PATH);
  const devin = createDevinProvider({ tokenStore, openBrowser: openUrlInBrowser });

  if (process.env.DEVIN_TOKEN) {
    console.error("Using DEVIN_TOKEN from the environment.");
    await devin.setToken(process.env.DEVIN_TOKEN);
  } else if (await tokenStore.get()) {
    console.error(`Using cached token from ${TOKEN_PATH}`);
  } else {
    console.error("No cached token found, starting Devin login...");
    await devin.login();
    console.error(`Login complete. Token cached at ${TOKEN_PATH}`);
  }

  console.error("Fetching available models...");
  const models = await devin.listModels();
  const model = models[0];
  if (!model) throw new Error("Devin returned no available models");
  console.error(`Using model ${model.id} (${model.name})`);

  console.error(`Sending: ${message}`);
  console.error("---");

  let sawText = false;
  for await (const event of devin.streamChat({
    model: model.id,
    systemPrompt: ["You are concise."],
    messages: [{ role: "user", content: message }]
  })) {
    if (event.type === "text_delta") {
      sawText = true;
      process.stdout.write(event.delta);
    } else if (event.type === "usage") {
      console.error(
        `\n[usage] input=${event.inputTokens} output=${event.outputTokens} cacheRead=${event.cacheReadTokens} cacheWrite=${event.cacheWriteTokens}`
      );
    } else if (event.type === "done") {
      if (sawText) process.stdout.write("\n");
      console.error(`[done] reason=${event.reason}`);
    }
  }
};

main().catch((error: unknown) => {
  if (error instanceof DevinAuthError) {
    console.error(`Devin login failed: ${error.message}`);
  } else if (error instanceof DevinApiError) {
    console.error(`Devin API request failed (${error.status}): ${error.message}`);
  } else if (error instanceof DevinProtocolError) {
    console.error(`Devin protocol error: ${error.message}`);
  } else {
    console.error(error);
  }
  process.exitCode = 1;
});

File: rust/examples/hello-world.rs
//! Manual smoke test for the Rust crate.
//!
//! Run:
//!   cargo run --manifest-path rust/Cargo.toml --example hello-world
//!   cargo run --manifest-path rust/Cargo.toml --example hello-world -- "Say hello."

use futures_util::StreamExt;
use std::{process::Command, sync::Arc};
use widevin::{
    DevinChatRequest, DevinContentPart, DevinError, DevinMessage, DevinProviderOptions,
    DevinStopReason, DevinStreamEvent, create_devin_provider, create_file_token_store,
};

const TOKEN_PATH: &str = ".devin-token";

fn open_browser(url: &str) {
    eprintln!("Open this URL to sign in to Devin: {url}");
    let result = if cfg!(target_os = "macos") {
        Command::new("open").arg(url).spawn()
    } else if cfg!(target_os = "windows") {
        Command::new("cmd").args(["/c", "start", "", url]).spawn()
    } else {
        Command::new("xdg-open").arg(url).spawn()
    };
    if result.is_err() {
        eprintln!("Could not launch a browser automatically; open the URL above manually.");
    }
}

#[tokio::main]
async fn main() -> Result<(), DevinError> {
    let message = std::env::args().skip(1).collect::<Vec<_>>().join(" ");
    let message = if message.is_empty() {
        "Say hello in one short sentence.".into()
    } else {
        message
    };
    let token_store = create_file_token_store(TOKEN_PATH);
    let provider = create_devin_provider(DevinProviderOptions {
        token_store: Some(token_store.clone()),
        open_browser: Some(Arc::new(|url| {
            Box::pin(async move {
                open_browser(&url);
                Ok(())
            })
        })),
        ..Default::default()
    });

    if let Ok(token) = std::env::var("DEVIN_TOKEN") {
        eprintln!("Using DEVIN_TOKEN from the environment.");
        provider.set_token(token).await?;
    } else if token_store.get().await?.is_some() {
        eprintln!("Using cached token from {TOKEN_PATH}");
    } else {
        eprintln!("No cached token found, starting Devin login...");
        provider.login().await?;
        eprintln!("Login complete. Token cached at {TOKEN_PATH}");
    }

    eprintln!("Fetching available models...");
    let model = provider
        .list_models()
        .await?
        .into_iter()
        .next()
        .ok_or_else(|| DevinError::protocol("Devin returned no available models"))?;
    eprintln!("Using model {} ({})", model.id, model.name);
    eprintln!("Sending: {message}");
    eprintln!("---");

    let mut events = provider.stream_chat(DevinChatRequest {
        model: model.id,
        system_prompt: vec!["You are concise.".into()],
        messages: vec![DevinMessage::User {
            content: vec![DevinContentPart::Text { text: message }],
        }],
        ..Default::default()
    });
    let mut saw_text = false;
    while let Some(event) = events.next().await {
        match event? {
            DevinStreamEvent::TextDelta { delta } => {
                saw_text = true;
                print!("{delta}");
            }
            DevinStreamEvent::Usage {
                input_tokens,
                output_tokens,
                cache_read_tokens,
                cache_write_tokens,
            } => eprintln!(
                "\n[usage] input={input_tokens} output={output_tokens} \
                 cacheRead={cache_read_tokens} cacheWrite={cache_write_tokens}"
            ),
            DevinStreamEvent::Done { reason } => {
                if saw_text {
                    println!();
                }
                let reason = match reason {
                    DevinStopReason::Stop => "stop",
                    DevinStopReason::Length => "length",
                    DevinStopReason::ToolUse => "toolUse",
                };
                eprintln!("[done] reason={reason}");
            }
            _ => {}
        }
    }
    Ok(())
}
